arangod/Aql/IndexNodeOptimizerRules.cpp

```diff
@@ -122,20 +125,27 @@ void arangodb::aql::lateDocumentMaterializationRule(Optimizer* opt,
       if (!indexNode->canApplyLateDocumentMaterializationRule() || indexNode->isLateMaterialized()) {
         continue; // loop is already optimized
       }
+      auto const* var = indexNode->outVariable();
+      std::vector<latematerialized::NodeExpressionWithAttrs> nodesToChange;
+      auto commonIndexId = IndexId::none();  // use one index only
       if (indexNode->hasFilter()) {
-        // IndexNode has an early pruning filter. In this case we cannot perform the optimization
-        continue;
+        auto* filter = indexNode->filter();
+        arangodb::containers::HashSet<Variable const*> currentUsedVars;
+        Ast::getReferencedVariables(filter->node(), currentUsedVars);
+        if (currentUsedVars.find(var) != currentUsedVars.end() &&
+            !processCalculationNode(indexNode, nullptr, filter, nodesToChange, commonIndexId)) {
+          // IndexNode has an early pruning filter which references not to index variables.
+          // In this case we cannot perform the optimization
+          continue;
+        }
       }
-      auto current = limitNode->getFirstDependency();
+      auto* current = limitNode->getFirstDependency();
```

Comment:
check current against nullptr ?

Reply:
I have just replaced other auto with auto*. There are many places. But nullptr cannot be there. I added TRI_ASSERT().